[WIP]Release unallocated reusable CPUs in reconcile loop

Chunxia202410 opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
In some cases, the app container can not reuse the CPU of InitContainer, which lead to CPU leaking.
Detail case refer to description in https://github.com/kubernetes/kubernetes/pull/131966 .

This PR implement the reconcile loop solution which is discussed with @ffromani in ( https://github.com/kubernetes/kubernetes/issues/112228#issuecomment-3066751798 and https://github.com/kubernetes/kubernetes/issues/112228#issuecomment-3088815625)

Below is the design overview:
In reconcile loop, for each pod (after all of the non-restartable initContainer terminated)
  - Calculate the leaked CPUs and release them.
  - Loop all non-restartable InitContainers (for pod restart case)
     - If InitContainer have leaked CPUs, reallocate CPUs for the InitContainer in pod other assigned CPUs (can’t use the CPUs of side-car container that runs before it).

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/112228

#### Special notes for your reviewer:
Need to consider Pod restart case, when Pod restart, the non-restartable InitContainer will restart, and need exclusive CPUs.

Chunxia202410 commented:
/sig node

Chunxia202410 commented:
When I test the Pod restart case, the non-restartable InitContainer will restart, and will not reallocate CPUs again. if no exclusive CPUs, it will use default CPUs. So this solution may not cover this case.

ffromani commented on pkg/kubelet/allocation/allocation_manager.go:
```diff
@@ -532,10 +546,8 @@ func allocationFromPod(pod *v1.Pod) state.PodResourceInfo {
-			alloc := *container.Resources.DeepCopy()
-			podAlloc.ContainerResources[container.Name] = alloc
-		}
+		alloc := *container.Resources.DeepCopy()
```
why is this done unconditionally now?

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -405,12 +406,49 @@ func (m *manager) removeStaleState() {
+			klog.V(5).InfoS("releasePodUnallocatedCPUs: skipping pod; status not found", "pod", klog.KObj(pod))
+			continue
+		}
+		for _, container := range pod.Spec.Containers {
+			cstatus, err := findContainerStatusByName(&pstatus, container.Name)
+			if err != nil {
+				klog.V(5).InfoS("releasePodUnallocatedCPUs: skipping container; container status not found in pod status", "pod", klog.KObj(pod), "containerName", container.Name, "err", err)
+				continue
+			}
+
+			if cstatus.State.Running != nil {
+				runningPod = true
+				break
+			}
+		}
```
I wonder if we should check 
1. if there's at least a app container running or
2. if all the non-sidecar init container are terminated

Did you evaluate both approaches?

Chunxia202410 replied:
Then I will judge the status of InitContainer one by one, so that it should be more safer

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -314,6 +316,41 @@ func (p *staticPolicy) updateCPUsToReuse(pod *v1.Pod, container *v1.Container, c
+	}
+}
+
+func (p *staticPolicy) inUseCPUsForNonRestartableInitContainer(s state.State, pod *v1.Pod, container *v1.Container) cpuset.CPUSet {
+	var behindContainer = false
+	inUseCPUs := cpuset.New()
+
+	for _, initContainer := range pod.Spec.InitContainers {
+		if container.Name == initContainer.Name {
+			if podutil.IsRestartableInitContainer(&initContainer) {
+				// If the container is a restartable init container, we should not
+				// reuse its cpuset, as a restartable init container can run with
+				// regular containers.
+				break
+			} else {
+				behindContainer = true
+				continue
+			}
+		}
+		if behindContainer == true && podutil.IsRestartableInitContainer(&initContainer){
+			cset, _ := s.GetCPUSet(string(pod.UID), initContainer.Name)
+			inUseCPUs = inUseCPUs.Union(cset)
+		}
+	}
+	for _, c := range pod.Spec.Containers {
+		cset, _ := s.GetCPUSet(string(pod.UID), c.Name)
+		inUseCPUs = inUseCPUs.Union(cset)
+	}
+	
+	return inUseCPUs
+}
```
here we want to find the cpus which were assigned to init container and are not transferred to app containers?
IOW, the intent is to find the CPUs leaked during the handoff between init and app containers?

ffromani reviewed (commented):
initial review

ffromani commented:
>Potential problem of this solution
>     When app container or side-car container resize, the initContainer will reallocate CPUs again even though it not >restart, and release in next reconcile loop.

I'm not yet up to speed with the container resizing/VPA so let me ask.
Why on resize of a running app or sidecar container the initContainers, which should be all terminated, will reallocate CPUs again? Who's reallocating them?

>    Non-restartable InitContainer will not be restricted resize anymore.

I'm not following this scenario, could you please elaborate a bit more?

ffromani reviewed (commented):
It's worth to add test(s), perhaps e2e, to exercise the fix so we can attempt a simpler change. What's the simpler reproducer for the reusable CPU leak? Let's have the test(s) in place.

Chunxia202410 commented on pkg/kubelet/allocation/state/state.go:
```diff
@@ -55,6 +55,7 @@ type Reader interface {
 
 type writer interface {
 	SetContainerResources(podUID types.UID, containerName string, resources v1.ResourceRequirements) error
+	SetContainerCPUResources(podUID types.UID, containerName string, request int64) error
```
I'm going to update InitContainer's CPU while reconcile, so we don't have to add this API.

Chunxia202410 committed:
Release leaked CPUs in reconcile loop

Add e2e case and UT case, four E2E test case mapping to
https://github.com/kubernetes/kubernetes/pull/131764#issuecomment-3431019564

# Conflicts:
#	pkg/kubelet/cm/cpumanager/cpu_manager.go
#	test/e2e_node/cpu_manager_test.go

Chunxia202410 commented:
> I was thinking something different, it seems.
> 
> The fundamental premise is we leak CPUs when init containers (any of them) ask for more exclusive CPUs than app container do. The extra CPUs unnecessary for long-running app or sidecar containers are kept busy and released only when the full pod terminates. Is my recollection correct or are there other source of leakage?
> 
> Internally, we record all the exclusive CPUs in the `cpusToReuse` data structure. We should pick from there CPUs needed by app containers (and perhaps sidecars? need to check), and remove from the `cpusToReuse` the cpus which we, indeed, reallocate to long-running containers. If my memory is not correct, please keep me honest
> 
> There's probably code which flushes `cpusToReuse` such only data for the last admitted pod is retained; but we can change the code to keep entries until they are removed **from the reconcile loop**, see below.
> 
> Once all init container are completed and all the app/sidecar containers are started, we therefore can have some stale entries in the `cpusToReuse` map, which are then the CPUs which we unnecessarily keep busy ("leaked") and which we can return to the shared pool.
> 
> So, in the reconcile loop, we can add a step to iterate over containers (or pods, fetching the pods through the containers, as needed), access the `cpusToReuse`, find the stale CPUs and release them.
> 
> If memory serves and the above is correct, then this is the reason why I was advocating this approach: it seems sufficient, self contained and require minimal changes; and, also a big plus, doesn't require the code to figure out which CPUs are leaked and we need to retrieve: we already have a detailed and precise record which we "just" need to access and act upon.
> 
> Let's verify carefully if my recollection and assumptions above are correct, please help me with that.
> 

I completely agree with what you're saying, and I can confirm that what you mentioned is correct.

> [EDIT] The (very nice, kudos!) diagram seems to be close to what I'm describing, but I'm somehow confused by the algo steps outlined above? Perhaps we are converging to the same end result from different perspectives?

However, when I consider the case of pod restarting, it will makes things complicated complex.
If we don't consider this case, I think what you've said is sufficient. 
But if we take this case into account, we need to consider reallocating CPU resources for the InitContainer, which might require adding some additional complex operations. 

For this design, I not release all the CPUs that the app or side-car container not used, because I think for the pod, some CPUs are free at pod running stage, but useful at pod restart stage. For example, in Case 4: No leaked CPUs, CPU 11 not used by app container, but it will used by InitContainer when pod restart, so I not release it.

What do you think about this case, or where should the code for reallocating CPU resources for the InitContainer be placed?

Chunxia202410 commented:
> ok, so the in-place resizing is what I was missing. Let me think about it.

Sorry, I write the wrong words, I want to say, pod restart.
But the pod resize also need to consider.

Chunxia202410 commented:
> > In fact, our design has also considered this problem, so we determine whether there are leaked CPUs by comparing the assigned CPU and the needed CPU number of the pod, rather than release all CPUs in cpusToReuse.
> 

Here the "needed CPU number of the pod" include the InitContainer CPUs, it calculated by function `func (p *staticPolicy) podGuaranteedCPUs`.

> This is however a generic concern. If the init container shape (= sum of resource requests) is bigger than the app (+sidecar) container shape, then by definition we will have cpu leakage (= cpus unnecessarily held because app containers don't need them), but then _if_ we release them, regardless of how we do that, then if pod restarts and init container need to run again, then the init containers won't have all the necessary resources unless we go through admission again, which I don't remember we do, but IIRC we don't (and seems pretty undesirable to redo admission). This is because cpu allocation only happens at admission time, which is already past and won't happen again.
> 
> Therefore, this concern is not about my algo vs your algo, but rather a generic problem about "should we release these CPUs at all or not"

If the init container shape (= sum of resource requests) is bigger than the app (+sidecar) container shape, for example, the sum of CPU resource requests of all init containers is 3, and the sum of CPU resource requests of all app containers is 2, the maximum of requested CPUs for the pod is 3. Please refer to the function `podGuaranteedCPUs` and https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers#resources-calculation-for-scheduling-and-pod-admission

So, in my opinion, if the pod assigned CPUs (Get by function `getPodAssignedCPUs`) is equal to the maximum of requested CPUs for the pod (Get by function `podGuaranteedCPUs`), there are no leaked CPUs. 
If  the pod assigned CPUs (Get by function `getPodAssignedCPUs`) is bigger than the maximum of requested CPUs for the pod (Get by function `podGuaranteedCPUs`), there are leaked CPUs, In this design, we release the leaked CPU, and reallocate CPUs for the InitContainer who have leaked CPUs, so that the InitContainer can running with enough CPUs when pod restart.

ffromani commented:
> > > If the init container shape (= sum of resource requests) is bigger than the app (+sidecar) container shape, for example, the sum of CPU resource requests of all init containers is 3, and the sum of CPU resource requests of all app containers is 2, the maximum of requested CPUs for the pod is 3. Please refer to the function `podGuaranteedCPUs` and https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers#resources-calculation-for-scheduling-and-pod-admission
> > 
> > 
> > Ok, the only thing to consider here is when sidecar container start, but other than that this is fine and matches our current conversation assumptions.
> 
> We never release the CPUs which used by the sidecar container, the CPUs will be removed from the `cpusToReuse` in function `updateCPUsToReuse`. The processing behavior of CPU assignments is the same for app container and sidecar.

Which is fine, as sidecar containers start alongside init containers but are long-running containers alongside app containers. The first part of the general problem is identifying the actually leaked CPUs, on which we agreed about the defintion few messages above, didn't we? (https://github.com/kubernetes/kubernetes/pull/131764#issuecomment-3479512434)

So it boils down to the init container (minus the sidecars) shape vs the long running container shape, and while the container restart is no concern, the pod restart can very much be, because the core issue is not going again through admission.
In-place pod downsize is another concern if brings us to the point on which the long-running (exclusive) CPU requirement is lower than init container CPU requirement, because this brings us to again to the pod restart concern.
Is this summarization right or are there more missing cases of concern?

> 
> > > So, in my opinion, if the pod assigned CPUs (Get by function `getPodAssignedCPUs`) is equal to the maximum of requested CPUs for the pod (Get by function `podGuaranteedCPUs`), there are no leaked CPUs. If the pod assigned CPUs (Get by function `getPodAssignedCPUs`) is bigger than the maximum of requested CPUs for the pod (Get by function `podGuaranteedCPUs`), there are leaked CPUs, In this design, we release the leaked CPU, and reallocate CPUs for the InitContainer who have leaked CPUs, so that the InitContainer can running with enough CPUs when pod restart.
> > 
> > 
> > I'll need to look carefully at the code again for both your designs
> 
> When I consider pod resize's case, when release leaked CPU in function `ReleaseLeakedCPUs`, it should not release CPUs directly from `cpusToReuse`, because the cpusToReuse not update when app/sidecar container scale down (function `updateCPUsToReuse` only can handle the case of pod create or pod scale up). It should recalculate the leaked CPUs in function `ReleaseLeakedCPUs`. I'll update this part of the code, and you can review the other code. thanks you~

Do we have already a cpumanager method that, given a currently running pod, identifies the leaked CPUs, possibly mapped to the init container which requested them?

Chunxia202410 commented:
> Which is fine, as sidecar containers start alongside init containers but are long-running containers alongside app containers. The first part of the general problem is identifying the actually leaked CPUs, on which we agreed about the defintion few messages above, didn't we? ([#131764 (comment)](https://github.com/kubernetes/kubernetes/pull/131764#issuecomment-3479512434))
> 

Yes.

> So it boils down to the init container (minus the sidecars) shape vs the long running container shape, and while the container restart is no concern, the pod restart can very much be, because the core issue is not going again through admission. In-place pod downsize is another concern if brings us to the point on which the long-running (exclusive) CPU requirement is lower than init container CPU requirement, because this brings us to again to the pod restart concern. Is this summarization right or are there more missing cases of concern?
> 

Yes, the summarization is right, and our design have covered this two cases.

> Do we have already a cpumanager method that, given a currently running pod, identifies the leaked CPUs, possibly mapped to the init container which requested them?

No, in current CPU manager, there are no parameters or function to record the leaked CPUs.

Chunxia202410 commented:
@ffromani , I update the code, I add a function `getPodNonUsedReusableCPUs` to get the nonAllocated reusable CPUs, PTAL thank you~

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -418,13 +419,61 @@ func (m *manager) removeStaleState(rootLogger logr.Logger) {
+		return false
+	}
+	for _, container := range pod.Spec.InitContainers {
+		if !podutil.IsRestartableInitContainer(&container) {
```
nit: let's try to keep the happy path on the left:
```
if podutil.IsRestartableInitContainer(&container) {
    continue
}
logger := klog.LoggerWithValues(podLogger, "containerName", container.Name)
cstatus, err := findContainerStatusByName(&pstatus, container.Name)
...
```

ffromani reviewed (commented):
Recently, we were discussing if restarts can raise the init containers to run again. If init container can run again under any circumstances, we cannot really release leaked CPUs because the init container will need them as requested. Did we reach a conclusion about the fact init container can ever need to run again?
